supervisely/api/volume/volume_annotation_api.py

```diff
@@ -197,24 +205,169 @@ def upload_paths(
-            self._api.volume.figure.upload_stl_meshes(
-                volume_id, ann.spatial_figures, key_id_map, interpolation_dir
-            )
             if progress_cb is not None:
                 progress_cb(1)
 
+    def _update_on_transfer(
+        self,
+        transfer_type: Literal["download", "upload"],
+        ann: VolumeAnnotation,
+        project_meta: ProjectMeta,
+        nrrd_paths: List[str],
+        key_id_map: KeyIdMap,
+        project_id: Optional[int] = None,
+    ) -> Tuple[VolumeAnnotation, ProjectMeta]:
+        """
+        Create new ObjClass and VolumeFigure annotations for converted STL.
+        Replace ClosedMeshSurface spatial figures with Mask3D.
+        Update the ann, project_meta, and key_id_map.
+
+        :param transfer_type: Defines the process during which the update will be performed ("download" or "upload").
+        :type transfer_type: Literal["download", "upload"]
+        :param ann: The VolumeAnnotation object to update.
+        :type ann: VolumeAnnotation
+        :param project_meta: The ProjectMeta object.
+        :type project_meta: ProjectMeta
+        :param nrrd_paths: Paths to the converted NRRD files from STL.
+        :type nrrd_paths: List[str]
+        :param key_id_map: The Key to ID map.
+        :type key_id_map: KeyIdMap
+        :param project_id: The Project ID to update metadata on upload (optional).
+        :type project_id: int, optional
+        :return: A tuple containing the updated ann and project_meta objects.
+        :rtype: Tuple[VolumeAnnotation, ProjectMeta]
+        """
+
+        def generate_unique_id(object_type: Literal["figures", "objects"]):
+            import random
+
+            entities_dict = key_id_map.to_dict().get(object_type)
+            if len(entities_dict) != 0:
+                max_id = max(entities_dict.values())
+                unique_id = max_id + (10 ** (len(str(max_id)) - 1)) + random.randint(100, 999)
+            else:
+                unique_id = 1
+            return unique_id
+
+        for nrrd_path in nrrd_paths:
+            object_key = None
+
+            # searching connection between interpolation and spatial figure in annotations and set its object_key
+            for sf in ann.spatial_figures:
+                if sf.key().hex == get_file_name(nrrd_path):
+                    object_key = sf.parent_object.key()
+                    break
+
+            if object_key:
+                for obj in ann.objects:
+                    if obj.key() == object_key:
+                        class_title = obj.obj_class.name
+                        break
+
+            else:
+                raise Exception(
+                    f"Can't find volume object for Mask3D from unknown file '{nrrd_path}'. Please check the project structure."
+                )
+
+            new_obj_class = ObjClass(f"{class_title}_mask_3d", Mask3D)
```

Comment:
cc = project_meta.get("{class_title}_mask_3d")
if cc is None:
      create new ccc

Reply:
+ flag